### ISSUE TITLE
Sign classification without GT label

### ISSUE THREAD

**Issue by matteorr** [leoshine/Spherical_Regression#2]
Congrats for the paper and thanks for the code! I have a question about how to use this method when there is no explicit GT for the sign(P).

Suppose that in the task of 3d pose estimation one is given the GT 3d pose in the world coordinate system and the CNN predicts the 3d pose in the camera coordinate system and a rotation matrix from camera to world coordinates. The loss would be computed as the L2 distance between the GT 3d pose and the predicted 3d pose after applying the predicted rotation. 

However, assigning the sign(P) using argmax to construct the rotation won't work since it is not differentiable. So it seems that the only way to train your classifier for sign is to have the direct GT for sign(P) (as you show in this [example](https://github.com/leoshine/Spherical_Regression/blob/master/code_logic.md)).

- Is this right?
- Any ideas on what a workaround might be?

Thanks a lot!

**Comment by leoshine** [comment]
Thanks for the comments. First of all, you are correct. 

I think an equivalent question is "how to use spherical exponentiation layer as an intermediate module where supervision is not available". It is true that the non-differentiability of argmax in the sign(P) prediction limits the application like this. 

However, I think there might be few workaround by replacing  argmax with its differentiable counterparts. For example, we can use Structured argmax [1] in deterministic inference (like this paper), or Gumbel-Softmax [2] in stochastic inference.

[1] H. Peng, S. Thomson, and N. A. Smith. Backpropagating through structured argmax using a spigot. ACL, 2018.
[2] E. Jang, S. Gu, and B. Poole. Categorical reparameterization with gumbel-softmax. ICLR, 2016.

**Comment by matteorr** [comment]
Thanks for the reply! I actually started implementing a version with the gumbel-softmax, so it's nice to have some confirmation of the validity of the idea. I didn't know about the other reference, thanks for the pointer. Will post here if I have updates on a working implementation.

**Comment by matteorr** [comment]
I have a follow-up question on this:

Instead of having another (sequential Linear model)[https://github.com/leoshine/Spherical_Regression/blob/master/S3.3D_Rotation/regQuatNet/regQuatNet.py#L225] for the prediction of the sign of the quaternion components would it be possible to use the sign values of the components before the special softmax activation? Basically the only difference is that there is no `self.head_signcate_quat` variable in the model and the sign is the same sign as the output of `self.head_sqrdprob_quat(x)`.

For instance, have you tried training a model in which (lines 247-253)[https://github.com/leoshine/Spherical_Regression/blob/master/S3.3D_Rotation/regQuatNet/regQuatNet.py#L247-L253] are replaced with the following?:

```
# pre softmax activation
x_sqr_quat = self.maskout(self.head_sqrdprob_quat(x).view(batchsize, self.nr_cate, self.reg_n_D   ), label)
x_sgc_quat = torch.sign(x_sqr_quat)
x_sqr_quat = self.softmax(x_sqr_quat_pre)
```
Thanks!

**Comment by leoshine** [comment]
I don’t think this will work.

The raw output from maskout, or we call them the “logits" (before softmax activation), are indeed in range of [-inf, inf].
However, since all logits are exponentiated in softmax, being negative or positive doesn’t related to the signs we meant in this context.
In my opinion, only the relative values within the logits decide the final output softmax, the signs of logits doesn’t matter.

You can also have a look at the "safe softmax” implementation (e.g. https://jamesmccaffrey.wordpress.com/2016/03/04/the-max-trick-when-computing-softmax/)
which makes softmax computation to be numerically stable.
It subtracts the logits by their maximum value first before passing it into softmax.
This makes all logits to be negative, whereas the softmax output is unchanged.

Best,
Shuai




On Jul 29, 2019, at 5:46 AM, matteorr <notifications@github.com<mailto:notifications@github.com>> wrote:


I have a follow-up question on this:

Instead of having another (sequential Linear model)[https://github.com/leoshine/Spherical_Regression/blob/master/S3.3D_Rotation/regQuatNet/regQuatNet.py#L225] for the prediction of the sign of the quaternion components would it be possible to use the sign values of the components before the special softmax activation? Basically the only difference is that there is no self.head_signcate_quat variable in the model and the sign is the same sign as the output of self.head_sqrdprob_quat(x).

For instance, have you tried training a model in which (lines 247-253)[https://github.com/leoshine/Spherical_Regression/blob/master/S3.3D_Rotation/regQuatNet/regQuatNet.py#L247-L253] are replaced with the following?:

# pre softmax activation
x_sqr_quat = self.maskout(self.head_sqrdprob_quat(x).view(batchsize, self.nr_cate, self.reg_n_D   ), label)
x_sgc_quat = torch.sign(x_sqr_quat)
x_sqr_quat = self.softmax(x_sqr_quat_pre)


Thanks!
